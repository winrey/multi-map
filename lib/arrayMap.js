"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayMultiMap = void 0;
/**
 * 注意，这个没有顺序区分
 */
class ArrayMultiMap {
    constructor(keysLens) {
        this._size = 0;
        this._data = new Map();
        if (keysLens <= 0) {
            throw new TypeError('key lens should greater than one');
        }
        if (typeof keysLens !== 'number') {
            throw new TypeError('key lens should be a number');
        }
        this._keysLens = keysLens;
    }
    get [Symbol.toStringTag]() {
        return 'ArrayMultiMap';
    }
    *[Symbol.iterator]() {
        for (const result of this.entries()) {
            yield result;
        }
    }
    get size() {
        return this._size;
    }
    get keysLens() {
        return this._keysLens;
    }
    checkKeysArg(keys) {
        if (keys.length !== this._keysLens) {
            throw new Error(`The length of the keys shouldn't change. old: ${this._keysLens}, yours:${keys.length}`);
        }
    }
    clone() {
        const clone = new ArrayMultiMap(this.keysLens);
        for (const [k, v] of this.entries()) {
            clone.set(k, v);
        }
        return clone;
    }
    set(keys, data) {
        /* istanbul ignore if  */
        if (this._keysLens === undefined) {
            this._keysLens = keys.length;
        }
        this.checkKeysArg(keys);
        keys.reduce((last, key, i) => {
            const isTheLastOne = i === keys.length - 1;
            if (isTheLastOne) {
                last.set(key, data);
                return;
            }
            if (last.has(key)) {
                return last.get(key);
            }
            const map = new Map();
            last.set(key, map);
            return map;
        }, this._data);
        this._size += 1;
        return this;
    }
    get(keys, defaultVal) {
        this.checkKeysArg(keys);
        const exit = Symbol('not found');
        const val = keys.reduce((last, key, i) => {
            if (last === exit) {
                return exit;
            }
            last = last;
            if (!last.has(key)) {
                return exit;
            }
            return last.get(key);
        }, this._data);
        return val === exit ? defaultVal : val;
    }
    has(keys) {
        this.checkKeysArg(keys);
        const exit = Symbol('not found');
        const val = keys.reduce((last, key, i) => {
            if (last === exit) {
                return exit;
            }
            last = last;
            const isTheLastOne = i === keys.length - 1;
            if (isTheLastOne) {
                return last.has(key);
            }
            if (!last.has(key)) {
                return exit;
            }
            return last.get(key);
        }, this._data);
        return val === exit ? false : val;
    }
    delete(keys) {
        this.checkKeysArg(keys);
        const exit = Symbol('not found');
        const maps = [this._data];
        const result = keys.reduce((last, key, i) => {
            if (last === exit) {
                return exit;
            }
            last = last;
            const isTheLastOne = i === keys.length - 1;
            if (isTheLastOne) {
                return last.delete(key);
            }
            if (!last.has(key)) {
                return exit;
            }
            const next = last.get(key);
            maps.push(next);
            return next;
        }, this._data);
        if (result === exit) {
            return false;
        }
        // clear unnecessary map
        for (let i = keys.length - 1; i > 0; i--) {
            if (maps[i].size) {
                break;
            }
            // need clear
            maps[i - 1].delete(keys[i - 1]);
        }
        if (result) {
            this._size -= 1;
        }
        return result;
    }
    clear() {
        this._size = 0;
        this._data = new Map();
    }
    forEach(callbackfn, thisArg) {
        const foreachIfMap = (map, keys = []) => {
            if (keys.length < this._keysLens) {
                // 未检索到指定层次
                /* istanbul ignore if  */
                if (!(map instanceof Map)) {
                    throw new Error('Inner Map Structure Error');
                }
                map.forEach((v, k) => {
                    foreachIfMap(v, [...keys, k]);
                });
                return;
            }
            // 检索到指定层次
            callbackfn.apply(thisArg, [map, keys, this]);
        };
        foreachIfMap(this._data);
    }
    *entries() {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const that = this;
        const foreachIfMap = function* (map, keys = []) {
            if (keys.length < that._keysLens) {
                // 未检索到指定层次
                /* istanbul ignore if  */
                if (!(map instanceof Map)) {
                    throw new Error('Inner Map Structure Error');
                }
                for (const [k, v] of map) {
                    for (const result of foreachIfMap(v, [...keys, k])) {
                        yield result;
                    }
                }
                return;
            }
            // 检索到指定层次?
            yield [keys, map];
        };
        for (const result of foreachIfMap(this._data)) {
            yield result;
        }
    }
    *keys() {
        for (const result of this.entries()) {
            yield result[0];
        }
    }
    *values() {
        for (const result of this.entries()) {
            yield result[1];
        }
    }
}
exports.ArrayMultiMap = ArrayMultiMap;
