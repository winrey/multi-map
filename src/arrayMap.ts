/**
 * 注意，这个没有顺序区分
 */
export class ArrayMultiMap<TKeys extends any[] = any[], TValue = string> implements Map<TKeys, TValue> {
  private _size = 0;
  private _data = new Map<any, any>();
  private _leafValue = Symbol('leafValue');
  private _defaultFunc?: (keys: TKeys) => TValue;

  constructor(defaultFunc?: (keys: TKeys) => TValue) {
    this._defaultFunc = defaultFunc;
  }

  get [Symbol.toStringTag]() {
    return 'ArrayMultiMap';
  }

  *[Symbol.iterator](): IterableIterator<[TKeys, TValue]> {
    for (const result of this.entries()) {
      yield result;
    }
  }

  get size() {
    return this._size;
  }

  clone() {
    const clone = new ArrayMultiMap<TKeys, TValue>();
    for (const [k, v] of this.entries()) {
      clone.set(k, v);
    }
    return clone;
  }

  set(keys: TKeys, data: TValue) {
    const map = keys.reduce((last: Map<any, any>, key, i) => {
      if (last.has(key)) {
        return last.get(key) as Map<any, any>;
      }
      const map = new Map<any, any>();
      last.set(key, map);
      return map;
    }, this._data);
    if (!map.has(this._leafValue)) {
      this._size += 1;
    }
    map.set(this._leafValue, data);
    return this;
  }

  get(keys: TKeys, defaultVal?: TValue): TValue | undefined {
    const exit = Symbol('not found');
    const val = keys.reduce((last: Map<any, any> | symbol, key, i) => {
      if (last === exit) {
        return exit;
      }
      last = last as Map<any, any>;
      if (!last.has(key)) {
        return exit;
      }
      return last.get(key) as Map<any, any>;
    }, this._data);
    if (val === exit || !val.has(this._leafValue)) {
      const result = defaultVal ?? this._defaultFunc?.(keys);
      if (result !== undefined) {
        this.set(keys, result);
      }
      return result;
    }
    return val.get(this._leafValue);
  }

  has(keys: TKeys): boolean {
    const exit = Symbol('not found');
    const lastMap = keys.reduce((last: Map<any, any> | symbol, key, i) => {
      if (last === exit) {
        return exit;
      }
      last = last as Map<any, any>;
      if (!last.has(key)) {
        return exit;
      }
      return last.get(key) as Map<any, any>;
    }, this._data);
    if (lastMap === exit) {
      return false;
    }
    return lastMap.has(this._leafValue);
  }

  delete(keys: TKeys) {
    const exit = Symbol('not found');
    const routeStack: Map<any, any>[] = [this._data];
    const result = keys.reduce((last: Map<any, any> | symbol, key, i) => {
      if (last === exit) {
        return exit;
      }
      last = last as Map<any, any>;
      if (!last.has(key)) {
        return exit;
      }
      const next = last.get(key) as Map<any, any>;
      routeStack.push(next);
      return next;
    }, this._data);

    if (result === exit || !result.has(this._leafValue)) {
      return false;
    }

    result.delete(this._leafValue);
    this._size -= 1;

    // clear unnecessary map
    for (let i = routeStack.length - 1; i > 0; i--) {
      if (routeStack[i].size) {
        break;
      }
      // need clear
      routeStack[i - 1].delete(keys[i - 1]);
    }
    return true;
  }

  clear(): void {
    this._size = 0;
    this._data = new Map<any, any>();
  }

  forEach(callbackfn: (value: TValue, keys: TKeys, map: Map<TKeys, TValue>) => void, thisArg?: any) {
    const foreachIfMap = (map: Map<any, any>, keys: any[] = []) => {
      map.forEach((v, k) => {
        if (k === this._leafValue) {
          callbackfn.apply(thisArg ?? this, [v, [...keys] as TKeys, this]);
          return;
        }
        foreachIfMap(v, [...keys, k]);
      });
    };
    foreachIfMap(this._data);
  }

  *entries(): IterableIterator<[TKeys, TValue]> {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const that = this;
    const foreachIfMap = function* (map: Map<any, any>, keys: any[] = []): any {
      for (const [k, v] of map) {
        if (k === that._leafValue) {
          yield [[...keys], v];
          break;
        }
        for (const result of foreachIfMap(v, [...keys, k])) {
          yield result;
        }
      }
    };
    for (const result of foreachIfMap(this._data)) {
      yield result;
    }
  }

  *keys(): IterableIterator<TKeys> {
    for (const result of this.entries()) {
      yield result[0];
    }
  }

  *values(): IterableIterator<TValue> {
    for (const result of this.entries()) {
      yield result[1];
    }
  }
}
