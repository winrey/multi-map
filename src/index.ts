export * from './arrayMap';
export * from './objectMap';
