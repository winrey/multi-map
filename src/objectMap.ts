import { ArrayMultiMap } from "./arrayMap";

export type TOptions = {
  ignoredExtarField?: boolean
}

/**
 * 注意，这个没有顺序区分
 */
export class ObjectMultiMap <TKey extends Record<string, any> = Record<string, any>, TValue=string> implements Map<TKey, TValue> {

  private _keyIndex: string[]
  private _data: ArrayMultiMap<any[], TValue>
  private _opts: TOptions

  get [Symbol.toStringTag]() {
    return "ObjectMultiMap";
  }

  *[Symbol.iterator](): IterableIterator<[TKey, TValue]> {
    for (const result of this.entries()) {
      yield result
    }
  }

  get size () {
    return this._data.size
  }

  get keyIndex () {
    return this._keyIndex
  }

  constructor(keyIndex: Array<string> | Record<string, any>, opts: TOptions = {
    ignoredExtarField: false,
  }) {
    this._keyIndex = this.geneKeysIndex(keyIndex)
    this._data = new ArrayMultiMap<any[], TValue>(this._keyIndex.length)
    this._opts = opts
  }

  private geneKeysIndex(key?: Array<string> | Record<string, any>) {
    if (!key || !["object", "array"].includes(typeof key)) {
      throw new TypeError("keyIndex type invalid")
    }
    if (!(key instanceof Array)) {
      key = Object.keys(key)
    }
    if (!key.length) {
      throw new TypeError("keyIndex lens invalid")
    }
    return key as string[]
  }

  private checkAndParseKeys(key: TKey, ignoredExtarField?: boolean) {
    ignoredExtarField = ignoredExtarField ?? this._opts.ignoredExtarField
    if (typeof key !== "object" || key instanceof Array) {
      throw new TypeError("key should be an obj")
    }
    let kkey = Object.keys(key)
    this._keyIndex.forEach(v => {
      const filtered = kkey.filter(k => k != v)
      if (kkey.length === filtered.length) {
        throw new TypeError(`index field "${v}" not found`)
      }
      kkey = filtered
    })
    if (!ignoredExtarField && kkey.length) {
      throw new TypeError(`key have extra fields: ${JSON.stringify(kkey)}`);
    }
    return this._keyIndex.map(k => key[k]) 
  }

  private index2Keys(index: any[]) {
    const result: Record<string, any> = {}
    /* istanbul ignore if  */
    if (index.length !== this._keyIndex.length) {
      throw new Error("index lens not matched")
    }
    for (let i = 0; i < index.length; i++) {
      result[this._keyIndex[i]] = index[i]
    }
    return result as TKey
  }

  set(keys: TKey, data: TValue) {
    const k = this.checkAndParseKeys(keys)
    this._data.set(k, data)
    return this
  }

  get(keys: TKey, defaultVal?: TValue): TValue {
    const k = this.checkAndParseKeys(keys)
    return this._data.get(k, defaultVal)
  }

  has(keys: TKey): boolean {
    const k = this.checkAndParseKeys(keys)
    return this._data.has(k)
  }

  delete(keys: TKey) {
    const k = this.checkAndParseKeys(keys)
    return this._data.delete(k)
  }

  clear(): void {
    this._data.clear()
  }

  
  forEach(callbackfn: (value: TValue, keys: TKey, map: Map<TKey, TValue>) => void, thisArg?: any) {
    this._data.forEach((v, k) => {
      const keys = this.index2Keys(k)
      callbackfn.apply(thisArg, [v, keys, this])
    })
  }

  *entries(): IterableIterator<[TKey, TValue]> {
    for (const [k, v] of this._data.entries()) {
      const keys = this.index2Keys(k)
      yield [keys, v]
    }
  }

  *keys(): IterableIterator<TKey> {
    for (const result of this.entries()) {
      yield result[0]
    }
  }

  *values(): IterableIterator<TValue> {
    for (const result of this.entries()) {
      yield result[1]
    }
  }
}
