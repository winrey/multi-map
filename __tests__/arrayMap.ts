
import { ArrayMultiMap } from '../src/arrayMap';

it("could be created", () => {
  new ArrayMultiMap(1)
  new ArrayMultiMap(10)
})

it("should check input error", () => {
  expect(() => new ArrayMultiMap(-1)).toThrow()
  expect(() => new ArrayMultiMap(0)).toThrow()
  expect(() => new ArrayMultiMap("test" as any)).toThrow()
  expect(() => new ArrayMultiMap(undefined as any)).toThrow()
})

it("could be set and get", () => {
  const map = new ArrayMultiMap(1)
  expect(map.has(["4"])).toBeFalsy()
  map.set(["4"], "233")
  map.set(["6"], "999")
  map.set([8], "888")
  expect(map.has(["4"])).toBeTruthy()

  expect(map.has(["5"])).toBeFalsy()
  expect(map.has(["6"])).toBeTruthy()
  expect(map.has([8])).toBeTruthy()
  expect(map.get(["4"])).toBe("233")
  expect(map.get(["6"])).toBe("999")
  expect(map.get(["8"])).toBe(undefined)
  expect(map.get([8])).toBe("888")
  expect(map.get(["4"], "666")).toBe("233")
  expect(map.get(["5"])).toBe(undefined)
  expect(map.get(["5"], "666")).toBe("666")
  expect(map.get(["6"], "666")).toBe("999")


  map.set(["4"], "666")
  expect(map.get(["4"])).toBe("666")
})

it("could be set and get multi", () => {
  const map = new ArrayMultiMap(3)
  expect(map.has(["4", "5", "6"])).toBeFalsy()
  map.set(["4", "5", "6"], "233")
  map.set(["4", "5", "8"], "2333")
  map.set(["4", "6", "6"], "23333")
  map.set(["5", "5", "6"], "233333")
  expect(map.has(["4", "5", "6"])).toBeTruthy()
  expect(map.has(["4", "5", "8"])).toBeTruthy()
  expect(map.has(["4", "6", "6"])).toBeTruthy()
  expect(map.has(["5", "5", "6"])).toBeTruthy()
  expect(map.get(["4", "5", "6"])).toBe("233")
  expect(map.get(["4", "5", "8"])).toBe("2333")
  expect(map.get(["4", "6", "6"])).toBe("23333")
  expect(map.get(["5", "5", "6"])).toBe("233333")
  expect(map.get(["1", "1", "1"], "")).toBe("")
  expect(map.get(["4", "5", "7"], "")).toBe("")
  expect(map.get(["4", "4", "6"], "")).toBe("")
  expect(map.get(["3", "5", "6"], "")).toBe("")
  expect(map.has(["1", "1", "1"])).toBeFalsy()
  expect(map.has(["4", "5", "7"])).toBeFalsy()
  expect(map.has(["4", "4", "6"])).toBeFalsy()
  expect(map.has(["3", "5", "6"])).toBeFalsy()
})

it("should throw if lens doesn't match", () => {
  const map = new ArrayMultiMap(1)
  expect(() => map.set([], "")).toThrow()
  expect(() => map.set(["4", "5"], "")).toThrow()
  expect(() => map.get(["4", "5"])).toThrow()
  expect(() => map.has([])).toThrow()
  expect(() => map.has(["4", "5"])).toThrow()
  expect(() => map.delete([])).toThrow()
  expect(() => map.delete(["4", "5"])).toThrow()
})

it("could delete and clear", () => {
  const map = new ArrayMultiMap(3)
  expect(map.size).toBe(0)
  map.set(["4", "5", "6"], "233")
  map.set(["4", "5", "8"], "2333")
  map.set(["4", "6", "6"], "23333")
  map.set(["5", "5", "6"], "233333")

  expect(map.size).toBe(4)
  expect(map.get(["4", "5", "6"])).toBe("233")
  expect(map.has(["4", "5", "6"])).toBeTruthy()

  expect(map.delete(["4", "5", "9"])).toBeFalsy()
  expect(map.delete(["4", "5", "6"])).toBeTruthy()
  expect(map.delete(["4", "5", "9"])).toBeFalsy()
  
  expect(map.size).toBe(3)
  expect(map.has(["4", "5", "6"])).toBeFalsy()
  expect(map.get(["4", "5", "6"])).toBe(undefined)

  expect(map.has(["4", "5", "8"])).toBeTruthy()
  expect(map.has(["4", "6", "6"])).toBeTruthy()
  expect(map.has(["5", "5", "6"])).toBeTruthy()

  map.clear()
  expect(map.size).toBe(0)

  expect(map.has(["4", "5", "8"])).toBeFalsy()
  expect(map.has(["4", "6", "6"])).toBeFalsy()
  expect(map.has(["5", "5", "6"])).toBeFalsy()
})

it("could be foreach", () => {
  const map = new ArrayMultiMap(3)
  map.set(["4", "5", "6"], "233")
  map.set(["4", "5", "8"], "2333")
  map.set(["4", "6", "6"], "23333")
  map.set(["5", "5", "6"], "233333")

  map.forEach((v, k) => {
    expect(v).toMatchSnapshot()
    expect(k).toMatchSnapshot()
  })
})

it("could be iterable", () => {
  const map = new ArrayMultiMap(3)
  map.set(["4", "5", "6"], "233")
  map.set(["4", "5", "8"], "2333")
  map.set(["4", "6", "6"], "23333")
  map.set(["5", "5", "6"], "233333")

  for (const result of map) {
    expect(result).toMatchSnapshot()
  }
})

it("entries could be for iterable", () => {
  const map = new ArrayMultiMap(3)
  map.set(["4", "5", "6"], "233")
  map.set(["4", "5", "8"], "2333")
  map.set(["4", "6", "6"], "23333")
  map.set(["5", "5", "6"], "233333")

  expect(map.entries()).toMatchSnapshot()

  for (const result of map.entries()) {
    expect(result).toMatchSnapshot()
  }
})

it("keys could be for iterable", () => {
  const map = new ArrayMultiMap(3)
  map.set(["4", "5", "6"], "233")
  map.set(["4", "5", "8"], "2333")
  map.set(["4", "6", "6"], "23333")
  map.set(["5", "5", "6"], "233333")

  expect(map.keys()).toMatchSnapshot()

  for (const result of map.keys()) {
    expect(result).toMatchSnapshot()
  }
})

it("values could be for iterable", () => {
  const map = new ArrayMultiMap(3)
  map.set(["4", "5", "6"], "233")
  map.set(["4", "5", "8"], "2333")
  map.set(["4", "6", "6"], "23333")
  map.set(["5", "5", "6"], "233333")

  expect(map.values()).toMatchSnapshot()

  for (const result of map.values()) {
    expect(result).toMatchSnapshot()
  }
})


