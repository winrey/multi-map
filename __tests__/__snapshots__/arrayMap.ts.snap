// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`could be foreach 1`] = `"233"`;

exports[`could be foreach 2`] = `
Array [
  "4",
  "5",
  "6",
]
`;

exports[`could be foreach 3`] = `"2333"`;

exports[`could be foreach 4`] = `
Array [
  "4",
  "5",
  "8",
]
`;

exports[`could be foreach 5`] = `"23333"`;

exports[`could be foreach 6`] = `
Array [
  "4",
  "6",
  "6",
]
`;

exports[`could be foreach 7`] = `"233333"`;

exports[`could be foreach 8`] = `
Array [
  "5",
  "5",
  "6",
]
`;

exports[`could be iterable 1`] = `
Array [
  Array [
    "4",
    "5",
    "6",
  ],
  "233",
]
`;

exports[`could be iterable 2`] = `
Array [
  Array [
    "4",
    "5",
    "8",
  ],
  "2333",
]
`;

exports[`could be iterable 3`] = `
Array [
  Array [
    "4",
    "6",
    "6",
  ],
  "23333",
]
`;

exports[`could be iterable 4`] = `
Array [
  Array [
    "5",
    "5",
    "6",
  ],
  "233333",
]
`;

exports[`entries could be for iterable 1`] = `Object {}`;

exports[`entries could be for iterable 2`] = `
Array [
  Array [
    "4",
    "5",
    "6",
  ],
  "233",
]
`;

exports[`entries could be for iterable 3`] = `
Array [
  Array [
    "4",
    "5",
    "8",
  ],
  "2333",
]
`;

exports[`entries could be for iterable 4`] = `
Array [
  Array [
    "4",
    "6",
    "6",
  ],
  "23333",
]
`;

exports[`entries could be for iterable 5`] = `
Array [
  Array [
    "5",
    "5",
    "6",
  ],
  "233333",
]
`;

exports[`keys could be for iterable 1`] = `Object {}`;

exports[`keys could be for iterable 2`] = `
Array [
  "4",
  "5",
  "6",
]
`;

exports[`keys could be for iterable 3`] = `
Array [
  "4",
  "5",
  "8",
]
`;

exports[`keys could be for iterable 4`] = `
Array [
  "4",
  "6",
  "6",
]
`;

exports[`keys could be for iterable 5`] = `
Array [
  "5",
  "5",
  "6",
]
`;

exports[`values could be for iterable 1`] = `Object {}`;

exports[`values could be for iterable 2`] = `"233"`;

exports[`values could be for iterable 3`] = `"2333"`;

exports[`values could be for iterable 4`] = `"23333"`;

exports[`values could be for iterable 5`] = `"233333"`;
